File: leetcodes/190. Reverse Bits/reverseBits.java
public class Solution {
  // you need treat n as an unsigned value
  public int reverseBits(int n) {
      //unsigned -> after reverse, put 0 in the left
      /**
       * bit by bit :
       * create newInt = 0
       * iterate over 32bits ->
       *    newInt << 1 -> left shift
       *    newInt += n & 1 -> get last bit
       *    n >> 1  -> right shift
       *
       * return newInt >>> 0 -> put 0 in the left if needed
       */
    if (n == 0) return 0;
    int newInt = 0;
    for (int i = 1; i <=32 ; i++){
      newInt <<= 1;
      newInt += (n&1);
      n >>=1;
    }
    return newInt;
  }
}

File: leetcodes/190. Reverse Bits/reverseBits.py
class Solution:
    def reverseBits(self, n: int) -> int:
      '''
      new =n &1 and n>>1 the n then
      newN = 0 | new and newN <<1
      iterate over 32 times O(1) space O(32)-> O(1)
      '''
      ans = 0
      for i in range(1,33):
        ans = (ans << 1) + (n&1)
        n = n>>1
      return ans
